File: Compiler/src/main/kotlin/nodes/ExprNode.java
package nodes;

public class ExprNode {
    public ExprType type = ExprType.UNINITIALIZED;

    public int id = 0;

    public boolean bool_value = false; // Значение бул
    public double number_value = 0.0; // Число
    public String string_value = ""; // Строка

    // Вызов функции
    public String ident = "";
    public ExprSeqNode args = null;

    public ExprNode adjusted_expr = null; // Выражение в скобках

    public FieldListNode table_constructor = null; // Таблица

    public ExprNode first_operand = null; // Первый операнд и операнд для унарных
    public ExprNode second_operand = null; // Второй операнд

    public VarNode var = null; // Переменная

    public ExprNode next = null;

    public static ExprNode createNilExprNode() {
        ExprNode result = new ExprNode();
        result.id = Utils.getLastId();
        result.type = ExprType.NIL;
        return result;
    }

    public static ExprNode createBoolExprNode(boolean val) {
        ExprNode result = new ExprNode();
        result.id = Utils.getLastId();
        result.type = ExprType.BOOLEAN;
        result.bool_value = val;
        return result;
    }

    public static ExprNode createNumberExprNode(float val) {
        ExprNode result = new ExprNode();
        result.id = Utils.getLastId();
        result.type = ExprType.NUMBER;
        result.number_value = val;
        return result;
    }

    public static ExprNode createStringExprNode(String val) {
        ExprNode result = new ExprNode();
        result.id = Utils.getLastId();
        result.type = ExprType.STRING;
        result.string_value = val;
        return result;
    }

    public static ExprNode createVarArgExprNode() {
        ExprNode result = new ExprNode();
        result.id = Utils.getLastId();
        result.type = ExprType.VAR_ARG;
        return result;
    }

    public static ExprNode createVarExprNode(VarNode var) {
        ExprNode result = new ExprNode();
        result.id = Utils.getLastId();
        result.type = ExprType.VAR;
        result.var = var;
        return result;
    }

    public static ExprNode createFunctionCallExprNode(String ID, ExprSeqNode args) {
        ExprNode result = new ExprNode();
        result.id = Utils.getLastId();
        result.type = ExprType.FUNCTION_CALL;
        result.ident = ID;
        result.args = args;
        return result;
    }

    public static ExprNode createAdjustingExprNode(ExprNode expr) {
        ExprNode result = new ExprNode();
        result.id = Utils.getLastId();
        result.type = ExprType.ADJUST;
        result.adjusted_expr = expr;
        return result;
    }

    public static ExprNode createTableConstructorExprNode(FieldListNode table) {
        ExprNode result = new ExprNode();
        result.id = Utils.getLastId();
        result.type = ExprType.TABLE_CONSTRUCTOR;
        result.table_constructor = table;
        return result;
    }

    public static ExprNode createBinExprNode(ExprType type_node, ExprNode first_operand, ExprNode second_operand) {
        ExprNode result = new ExprNode();
        result.id = Utils.getLastId();
        result.type = type_node;
        result.first_operand = first_operand;
        result.second_operand = second_operand;
        return result;
    }

    public static ExprNode createUnaryExprNode(ExprType type_node, ExprNode operand) {
        ExprNode result = new ExprNode();
        result.id = Utils.getLastId();
        result.type = type_node;
        result.first_operand = operand;
        return result;
    }
}

File: Compiler/src/main/kotlin/nodes/ExprNode.kt
package nodes

class ExprNode {
    var type = ExprType.UNINITIALIZED
    var id = 0
    var boolValue = false // Значение бул
    var floatNumberValue = 0.0 // Число
    var intNumberValue = 0 // Число
    var stringValue = "" // Строка

    // Вызов функции
    var ident = ""
    var args: ExprSeqNode? = null
    var adjustedExpr: ExprNode? = null // Выражение в скобках
    var tableConstructor: FieldListNode? = null // Таблица
    var firstOperand: ExprNode? = null // Первый операнд и операнд для унарных
    var secondOperand: ExprNode? = null // Второй операнд
    var varNode: VarNode? = null // Переменная
    var next: ExprNode? = null

    companion object {
        fun createNilExprNode(): ExprNode {
            val result = ExprNode()
            result.id = Utils.lastId
            result.type = ExprType.NIL
            return result
        }

        fun createBoolExprNode(value: Boolean): ExprNode {
            val result = ExprNode()
            result.id = Utils.lastId
            result.type = ExprType.BOOLEAN
            result.boolValue = value
            return result
        }

        fun createFloatNumberExprNode(value: Double): ExprNode {
            val result = ExprNode()
            result.id = Utils.lastId
            result.type = ExprType.FLOAT_NUMBER
            result.floatNumberValue = value
            return result
        }

        fun createIntNumberExprNode(value: Int): ExprNode {
            val result = ExprNode()
            result.id = Utils.lastId
            result.type = ExprType.INT_NUMBER
            result.intNumberValue = value
            return result
        }

        fun createStringExprNode(value: String): ExprNode {
            val result = ExprNode()
            result.id = Utils.lastId
            result.type = ExprType.STRING
            result.stringValue = value
            return result
        }

        fun createVarArgExprNode(): ExprNode {
            val result = ExprNode()
            result.id = Utils.lastId
            result.type = ExprType.VAR_ARG
            return result
        }

        fun createVarExprNode(varNode: VarNode?): ExprNode {
            val result = ExprNode()
            result.id = Utils.lastId
            result.type = ExprType.VAR
            result.varNode = varNode
            return result
        }

        fun createFunctionCallExprNode(ID: String, args: ExprSeqNode?): ExprNode {
            val result = ExprNode()
            result.id = Utils.lastId
            result.type = ExprType.FUNCTION_CALL
            result.ident = ID
            result.args = args
            return result
        }

        fun createAdjustingExprNode(expr: ExprNode?): ExprNode {
            val result = ExprNode()
            result.id = Utils.lastId
            result.type = ExprType.ADJUST
            result.adjustedExpr = expr
            return result
        }

        fun createTableConstructorExprNode(table: FieldListNode?): ExprNode {
            val result = ExprNode()
            result.id = Utils.lastId
            result.type = ExprType.TABLE_CONSTRUCTOR
            result.tableConstructor = table
            return result
        }

        fun createBinExprNode(type_node: ExprType, first_operand: ExprNode?, second_operand: ExprNode?): ExprNode {
            val result = ExprNode()
            result.id = Utils.lastId
            result.type = type_node
            result.firstOperand = first_operand
            result.secondOperand = second_operand
            return result
        }

        fun createUnaryExprNode(type_node: ExprType, operand: ExprNode?): ExprNode {
            val result = ExprNode()
            result.id = Utils.lastId
            result.type = type_node
            result.firstOperand = operand
            return result
        }
    }
}